287 make better use of sidepanel  option to list playlists

abegert opened the pull request:
This implements https://github.com/xou816/spot/issues/287

I replaced the StackSideBar with a ListBox, so that we can freely modify it. I'm still using a Stack for the pages in the mainwindow and switch to them manually when the corresponding ListBoxRow is activated.

The UI looks like this now, the only change visually since the last discussion is the 'All Playlists' entry:
![image](https://user-images.githubusercontent.com/91668665/146778355-080fcdb1-a847-46bb-bf60-af5893c13835.png)

I'm happy with it, though I can think of one thing that could be worth improving: It would be good to differentiate the 'All Playlists' entry a little from the users playlists below. The only idea I've had was to add something like separators around it, but not sure if that's such a good idea.

Thoughts, @jannuary @xou816 ?

jannuary commented:
A design review:
- Sidebar items are now edge to edge, I don't think this is something we want?
![image](https://user-images.githubusercontent.com/27908024/146783501-302a77bd-ca36-44f6-b1c8-099d0bac8853.png)
FYI .navigation-sidebar on the ListBox gives proper paddings and rounded corners (though you'd need to apply ```background-color: @view_bg_color;``` in css manually)
- The sidebar seems wider than before? I think the status quo was mostly fine, I don't think playlist names are gonna be significantly long for the most part.
![image](https://user-images.githubusercontent.com/27908024/146784318-21e9b6aa-6b35-41bc-82ca-7ec8f3debcdb.png)
- 'All Playlists' has the same icon as every other playlist, despite not being a playlist. I think it'd help to delineate it with another icon
(though the problem there is what icon to use)

Other than that, great work! Would love to see this merged!

jannuary commented on src/app/components/sidebar_listbox/icons/star-filled-symbolic.svg:
```diff
@@ -0,0 +1,151 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This doesn't need to be bundled - just use `starred-symbolic`.

jannuary commented on src/app/components/sidebar_listbox/sidebar.css:
```diff
@@ -0,0 +1,3 @@
+.item_sidebar {
```
Do we need css here?

jannuary replied:
Without `padding`, of course

jannuary commented on src/app/components/sidebar_listbox/mod.rs:
```diff
@@ -0,0 +1,42 @@
+        let label = gtk::Label::new(Option::from(
+            item.property("title").unwrap().get::<&str>().unwrap(),
+        ));
+        let gtk_box = gtk::Box::new(gtk::Orientation::Horizontal, 10);
```
```suggestion
        let gtk_box = gtk::Box::new(gtk::Orientation::Horizontal, 12);
```

abegert commented:
@jannuary thanks for the prompt and detailed feedback!

* I added the css class 'navigation-sidebar' and set the background-color
* I didn't deliberately change the width, it would just auto expand to accommodate the playlist with the longest name. In the case of your screenshot to the width of the second playlist from the bottom. I just made a change that limits the label to 20 characters and ellipsizes after that. I also added a tooltip so longer names can still be read somehow
* Yeah, I wasn't happy about that either, but I just found an icon that I like for it. It's a grid symbol (see the screenshot below)

The rest of your suggested changes I implemented as well.

![image](https://user-images.githubusercontent.com/91668665/146799911-2b8d65e9-6ce5-4c76-9c11-646918075552.png)


Let me know what you think, I hope I understood your suggestions correctly!

jannuary commented:
I'm not too partial to the grid icon, might mess around with it tomorrow, but otherwise LGTM.

xou816 reviewed (commented):
Hi! This is astounding work, thank you for working on this! I have a few comments here and there (most importantly, I think we could avoid direct calls to `property` which are "flimsy" since one could easily get the wrong string here or mess up the conversion from a GSomething to a plain rust type) but for the most part the implementation is pretty good :) I'll try and keep a close eye on the PR if you need more feedback from me.

Also thanks @jannuary for helping with the review, I appreciate it!

abegert commented on src/app/components/navigation/factory.rs:
```diff
@@ -37,9 +37,12 @@ impl ScreenFactory {
         )
     }
 
+    pub fn make_saved_playlists_model(&self) -> SavedPlaylistsModel {
```
I can't make it non public without some modifcation: the method is called from HomePane as well. I need the SavedPlaylistModel there, and preferred just making a new one over changing the return type of make_saved_playlists.
I'd be open for a suggestion on how to change this, it's definitely not perfect.

abegert commented on src/app/components/navigation/navigation.rs:
```diff
@@ -11,7 +11,8 @@ pub struct Navigation {
     leaflet: libadwaita::Leaflet,
     navigation_stack: gtk::Stack,
-    home_stack_sidebar: gtk::StackSidebar,
+    home_listbox: gtk::ListBox,
```
I think we do. We pass it to the HomePane constructor in make_home. make_home is called from navigation, for example during push_screen.

abegert commented on src/app/components/sidebar_listbox/mod.rs:
```diff
@@ -0,0 +1,47 @@
+use gtk::ListBox;
+pub use sidebar_item::*;
+
+pub fn build_sidebar_listbox(builder: &gtk::Builder, list_store: &gio::ListStore) -> ListBox {
```
Do you mean a custom ListBox widget? I'm not sure whats to gain from that. I did refactor this function. It now just passes the item from the ListStore into the SideBarRow constructor, which then handles all the details.
I also wrote a custom widget for the GtkBox containing an icon and a label with certain styling, etc.

xou816 replied:
why a custom widget -> for consistency and easy editing (ui files) but thats already cleaner, I'm ok with the current solution :)

abegert commented on src/app/components/sidebar_listbox/mod.rs:
```diff
@@ -0,0 +1,47 @@
+        let gtk_box = gtk::Box::new(gtk::Orientation::Horizontal, 12);
+        gtk_box.set_tooltip_text(title);
+        let row = SideBarRow::new(item.property("id").unwrap().get::<&str>().unwrap());
+        if item.property("grayedout").unwrap().get::<bool>().unwrap() {
```
Good suggestion, I had tried to do that but failed casting the item into SideBarItem. I managed to do that this time, it's much cleaner like that imo.

abegert commented:
Thank you for the kind words! I worked in your feedback, let me know what you think.

xou816 commented on src/app/components/navigation/home.rs:
```diff
@@ -130,7 +121,7 @@ impl HomePane {
+        let model = self.saved_playlists_model.clone();
         self.listbox
             .connect_row_activated(clone!(@weak self.stack as stack => move |_, row| {
                 let n = row.property("id").expect("Could not get id of ListBoxRow.");
```
downcast + use your new id() method maybe?

xou816 commented on src/app/components/sidebar_listbox/mod.rs:
```diff
@@ -11,34 +12,9 @@ pub use sidebar_item::*;
-            row.set_child(Option::from(&gtk_box));
-        }
+    listbox.bind_model(Some(list_store), move |obj| {
+        let item = obj.clone().downcast::<sidebar_item::SideBarItem>().unwrap();
```
instead of clone+downcast I think you can just downcast_ref !

xou816 commented:
Think we're good to merge! Thanks a lot for this!!
